src/fsck/fsck.c

```diff
@@ -402,7 +401,7 @@ static int run(int argc, char *argv[]) {
         }
 
         progress_pipe[1] = safe_close(progress_pipe[1]);
-        (void) process_progress(TAKE_FD(progress_pipe[0]));
+        (void) process_progress(TAKE_FD(progress_pipe[0]), console);
```

Comment:
The problem is not that `console` is opened inside `process_progress`, but that `progress_pipe[0]` is closed unconditionally, even if the console is not available. My suggestion would be to change `process_progress` to take `int *`, and close the fd only if not returning early.
I.e. the diff would be something like this:
```diff
- (void) process_progress(TAKE_FD(progress_pipe[0]));
+ (void) process_progress(&progress_pipe[0]);
```
... and of course process_progress has to be adjusted too.

Reply:
OK, I see what you are trying to do. I was wrong, and your patch cannot cause a crash. But it still causes a leak: nothing closes `console`. Please consider reworking it as I suggested above.

Reply:
With the path `console` is declared in `run` function but still declared with `_cleanup_fclose_`. So `console` will be automatically closed at end of `run` function.
About the potential issue you see with `console == NULL`, This case cannot happen but I agree this is not very readable. I can still add a test like under:
```diff
-        progress_pipe[1] = safe_close(progress_pipe[1]);
-        (void) process_progress(TAKE_FD(progress_pipe[0]), console);
+        if (console) {
+                progress_pipe[1] = safe_close(progress_pipe[1]);
+                (void) process_progress(TAKE_FD(progress_pipe[0]), console);
+        }
```

Reply:
Thanks, that looks good.